File: compiler/compiler/src/main/scala/io/github/spaceman1701/oxidizer/compiler/bytecode/Instruction.scala
package io.github.spaceman1701.oxidizer.compiler.bytecode

import io.github.spaceman1701.oxidizer.compiler.util.{U32, U64, U16}

sealed trait Instruction {
  val opcode: Byte
  val size: Byte = 1
}

/*
Basic Load/Store instructions. For simplicity only wide insturctions.
 */
case class Store(loc: U16) extends Instruction { //objref | value -> [] (convert to objref and place on data stack)
  val opcode = 0
}
case class LoadConstInt(value: Long) extends Instruction {
  val opcode = 1
}
case class LoadConstFloat(value: Double) extends Instruction {
  val opcode = 2
}
case class LoadConstStr(ptr: U32) extends Instruction { //load a string from the constant string areas
  val opcode = 3
}
case class LoadVar(loc: U16) extends Instruction {
  val opcode = 4
}

case object CreateObject extends Instruction { // objref(name) -> objref
  val opcode = 5
}

/*
Native list slice. Allows complex slices to be done in one instruction
 */
case object SliceList extends Instruction { //objref, objref -> objref
  val opcode = 6
}

/*
Math instructions. Short circuit to native impl when available
 */
//objref, objref -> value
case object Add extends Instruction {
  val opcode = 7
}
case object Sub extends Instruction {
  val opcode = 8
}
case object Mul extends Instruction{
  val opcode = 9
}
case object Div extends Instruction{
  val opcode = 10
}
case object Mod extends Instruction{
  val opcode = 11
}
case object Concat extends Instruction{
  val opcode = 12
}
case object Pow extends Instruction{
  val opcode = 13
}

case object LAnd extends Instruction {
  val opcode = 14
}

case object LOr extends Instruction {
  val opcode = 15
}

/*
Compare instructions. For when result of a compare is stored
 */
case object CompG extends Instruction { //objref, objref -> value
  val opcode = 16
}
case object CompL extends Instruction {
  val opcode = 17
}
case object CompEq extends Instruction {
  val opcode = 18
}

/*
Load class member (function or field)
 */
case object LoadMember extends Instruction { //objref, name -> objref
  val opcode = 19
}

/*
Store top into pointer at top-1.

i.e
LoadVar(0) //objref
LoadConstString(foo) //objref, name
LoadMember //objref
LoadConstInt(0) //objref, value
StoreMember //[] (the objref now points to value)

 */
case object StoreMember extends Instruction { //objref, objref -> []
  val opcode = 20
}

/*
Call a function. Creates a new data stack and a new execution stack
The tos should be a callable, the next value should be a param count, and the next values
should be n params
 */
//objref(callable), objref(int), objref... -> []
case object Call extends Instruction {
  val opcode = 21
}


//unconditional jump
case class Jump(target: U32) extends Instruction {
  val opcode = 22
}

/*
Branch instrictions. Normal and wide as most branches are very small. Jumps can only occur within a single module
so wide instructions are safe as unsigned Int
 */
//objref, objref -> []
//jumps can only happen inside a file
//thus, hard limit on 4,294,967,295 per file
case class IfFalse(target: U32) extends Instruction {
  val opcode = 23
}


case object Ret extends Instruction {
  val opcode = 24
}


case object Throw extends Instruction { //objref(exeception) -> [] //clears the execution stack
  val opcode = 25
}

/*
Async IO instructions.
 */
case object PostEvent extends Instruction { //objref(event) -> objref-promise
  val opcode = 26
}
case object WaitEvent extends Instruction { //objref-promise | objref -> objref
  val opcode = 27
}

/*
Spawn new coroutine with a function
 */
case object SpawnCoroutine extends Instruction { //objref(callable)
  val opcode = 28
}

/*
IPC instructions. Write suspends current thread. Read suspends until object is placed on channel
 */
//TODO: Maybe these should use PostEvent and WaitEvent
case object WriteChannel extends Instruction { //objref -> []
  val opcode = 29
}
case object ReadChannel extends Instruction { //objref(channel) -> objref
  val opcode = 30
}

case object Not extends Instruction {
  val opcode = 31
}

case object  XOr extends Instruction {
  val opcode = 32
}

case object BAnd extends Instruction {
  val opcode = 33
}

case object  BOr extends Instruction {
  val opcode = 34
}

case object LeftShift extends Instruction {
  val opcode = 35
}

case object RightShift extends Instruction {
  val opcode = 36
}

case object URightShift extends Instruction {
  val opcode = 37
}

case object Modulo extends Instruction {
  val opcode = 38
}

case object BCompliment extends Instruction {
  val opcode = 39
}

case object Dup extends Instruction { //objref -> objref, objref
  val opcode = 40
}
/*
Following instructions:
Used for the bytecode generator. Cannot be produced in real code.
 */
case object NoOp extends Instruction {
  val opcode: Byte = -1
}

case class Break(loopContext: LoopContext) extends Instruction {
  val opcode: Byte = -2
}

case class Continue(loopContext: LoopContext) extends Instruction {
  val opcode: Byte = -3
}

File: oxidizer-linux/src/instruction.rs
use std::mem;
use crate::instruction::Instruction::*;
pub enum Instruction {
    Store(u16),
    LoadConstInt(i64),
    LoadConstFloat(f64),
    LoadConstStr(u32),
    LoadVar(u16),
    CreateObject,

    SliceList,

    Add,
    Sub,
    Mul,
    Div,
    Mod,
    Concat,
    Pow,
    LAnd,
    LOr,

    CompG,
    CompL,
    CompEq,

    LoadMember,
    StoreMember,

    Call,

    Jump(u32),
    IfFalse(u32),

    Ret,

    Throw,

    PostEvent,
    WaitEvent,

    Spawn,

    WriteChannel,
    ReadChannel,

    Not,
    XOr,
    BAnd,
    BOr,
    LeftShift,
    RightShift,
    URightShift,
    Modulo,
    BCompliment,

    Dup,
    Consume,

    SwapTOS2WithTOS3
}

impl Instruction {
    pub fn decode(bytes: &[u8]) -> Option<(Instruction, usize)> {
        match bytes[0] {
            0x0 => Some((Store(Instruction::param_to_u16(bytes)), 3 as usize)),
            0x1 => Some((LoadConstInt(Instruction::param_to_i64(bytes)), 9 as usize)),
            0x2 => Some((LoadConstFloat(Instruction::param_to_f64(bytes)), 9 as usize)),
            0x3 => Some((LoadConstStr(Instruction::param_to_u32(bytes)), 5 as usize)),
            0x4 => Some((LoadVar(Instruction::param_to_u16(bytes)), 3 as usize)),
            0x5 => Some((CreateObject, 1 as usize)),
            0x6 => Some((SliceList, 1 as usize)),
            0x7 => Some((Add, 1 as usize)),
            0x8 => Some((Sub, 1 as usize)),
            0x9 => Some((Mul, 1 as usize)),
            0xA => Some((Div, 1 as usize)),
            0xB => Some((Mod, 1 as usize)),
            0xC => Some((Concat, 1 as usize)),
            0xD => Some((Pow, 1 as usize)),
            0xE => Some((LAnd, 1 as usize)),
            0xF => Some((LOr, 1 as usize)),
            0x10 => Some((CompG, 1 as usize)),
            0x11 => Some((CompL, 1 as usize)),
            0x12 => Some((CompEq, 1 as usize)),
            0x13 => Some((LoadMember, 1 as usize)),
            0x14 => Some((StoreMember, 1 as usize)),
            0x15 => Some((Call, 1 as usize)),
            0x16 => Some((Jump(Instruction::param_to_u32(bytes)), 5 as usize)),
            0x17 => Some((IfFalse(Instruction::param_to_u32(bytes)), 5 as usize)),
            0x18 => Some((Ret, 1 as usize)),
            0x19 => Some((Throw, 1 as usize)),
            0x1A => Some((PostEvent, 1 as usize)),
            0x1B => Some((WaitEvent, 1 as usize)),
            0x1C => Some((Spawn, 1 as usize)),
            0x1D => Some((WriteChannel, 1 as usize)),
            0x1E => Some((ReadChannel, 1 as usize)),
            0x1F => Some((Not, 1 as usize)),
            0x20 => Some((XOr, 1 as usize)),
            0x21 => Some((BAnd, 1 as usize)),
            0x22 => Some((BOr, 1 as usize)),
            0x23 => Some((LeftShift, 1 as usize)),
            0x24 => Some((RightShift, 1 as usize)),
            0x25 => Some((URightShift, 1 as usize)),
            0x26 => Some((Modulo, 1 as usize)),
            0x27 => Some((BCompliment, 1 as usize)),
            0x28 => Some((Dup, 1 as usize)),
            0x29 => Some((Consume, 1 as usize)),
            0x2A => Some((SwapTOS2WithTOS3, 1 as usize)),
            _ => {
                println!("bad opcode {}", bytes[0]);
                None
            }
        }
    }

    fn param_to_u16(data: &[u8]) -> u16 {
        let sized_data = [data[1], data[2]];
        unsafe {
            mem::transmute(sized_data)
        }
    }

    fn param_to_u32(data: &[u8]) -> u32 {
        let sized_data = [data[1], data[2], data[3], data[4]];
        let foo = unsafe {
            mem::transmute(sized_data)
        };
        foo
    }

    fn param_to_i64(data: &[u8]) -> i64 {
        let sized_data = [data[1], data[2], data[3], data[4], data[5], data[6], data[7], data[8]];
        unsafe {
            mem::transmute(sized_data)
        }
    }


    fn param_to_f64(data: &[u8]) -> f64 {
        let sized_data = [data[1], data[2], data[3], data[4], data[5], data[6], data[7], data[8]];
        unsafe {
            mem::transmute(sized_data)
        }
    }
}
